src/analysis/calculus/iterated_deriv.lean

```diff
@@ -0,0 +1,296 @@
+by { ext x, refl }
+
+/-- Write the iterated Fréchet derivative as the composition of a continuous linear equiv and the
+iterated derivative. -/
+lemma iterated_fderiv_eq_equiv_comp :
+  iterated_fderiv 𝕜 n f
+  = (continuous_multilinear_map.pi_field_equiv 𝕜 (fin n) F) ∘ (iterated_deriv n f) :=
+begin
+  rw [iterated_deriv_eq_equiv_comp, ← function.comp.assoc,
+      continuous_linear_equiv.self_comp_symm],
+  refl
+end
+
+/-- The `n`-th Fréchet derivative applied to a vector `(m 0, ..., m (n-1))` is the derivative
+multiplied by the product of the `m i`s. -/
+lemma iterated_fderiv_apply_eq_iterated_deriv_mul_prod {m : (fin n) → 𝕜} :
+  (iterated_fderiv 𝕜 n f x : ((fin n) → 𝕜) → F) m = finset.univ.prod m • iterated_deriv n f x :=
+by { rw [iterated_deriv_eq_iterated_fderiv, ← continuous_multilinear_map.map_smul_univ], simp }
+
+@[simp] lemma iterated_deriv_zero :
+  iterated_deriv 0 f = f :=
+by { ext x, simp [iterated_deriv] }
+
+@[simp] lemma iterated_deriv_one :
+  iterated_deriv 1 f = deriv f :=
+by { ext x, simp [iterated_deriv], refl }
+
+/-- The property of being `C^n`, initially defined in terms of the Fréchet derivative, can be
+reformulated in terms of the one-dimensional derivative. -/
+lemma times_cont_diff_iff_iterated_deriv {n : with_top ℕ} :
+  times_cont_diff 𝕜 n f ↔
+(∀m:ℕ, (m : with_top ℕ) ≤ n → continuous (iterated_deriv m f))
+∧ (∀m:ℕ, (m : with_top ℕ) < n → differentiable 𝕜 (iterated_deriv m f)) :=
+by simp only [times_cont_diff_iff_continuous_differentiable, iterated_fderiv_eq_equiv_comp,
+  continuous_linear_equiv.comp_continuous_iff,
+  continuous_linear_equiv.comp_differentiable_iff]
+
+/-- To check that a function is `n` times continuously differentiable, it suffices to check that its
+first `n` derivatives are differentiable. This is slightly too strong as the condition we
+require on the `n`-th derivative is differentiability instead of continuity, but it has the
+advantage of avoiding the discussion of continuity in the proof (and for `n = ∞` this is optimal).
+-/
+lemma times_cont_diff_of_differentiable_iterated_deriv {n : with_top ℕ}
+  (h : ∀(m : ℕ), (m : with_top ℕ) ≤ n → differentiable 𝕜 (iterated_deriv m f)) :
+  times_cont_diff 𝕜 n f :=
+times_cont_diff_iff_iterated_deriv.2
+  ⟨λ m hm, (h m hm).continuous, λ m hm, (h m (le_of_lt hm))⟩
+
+lemma times_cont_diff.continuous_iterated_deriv {n : with_top ℕ} (m : ℕ)
+  (h : times_cont_diff 𝕜 n f) (hmn : (m : with_top ℕ) ≤ n) :
+  continuous (iterated_deriv m f) :=
+(times_cont_diff_iff_iterated_deriv.1 h).1 m hmn
+
+lemma times_cont_diff.differentiable_iterated_deriv {n : with_top ℕ} (m : ℕ)
+  (h : times_cont_diff 𝕜 n f) (hmn : (m : with_top ℕ) < n) :
+  differentiable 𝕜 (iterated_deriv m f) :=
+(times_cont_diff_iff_iterated_deriv.1 h).2 m hmn
+
+/-- The `n+1`-th iterated derivative can be obtained by differentiating the `n`-th
+iterated derivative. -/
+lemma iterated_deriv_succ : iterated_deriv n.succ f = deriv (iterated_deriv n f) :=
```

Comment:
I wonder if it is better to write `n+1` rather than `n.succ`. I’m not sure if there is a rule about this, but I find rewriting back and forth between these tedious.

Reply:
Done, in this file and `times_cont_diff.lean`.